witchcraft/server_metrics.go

```diff
@@ -118,6 +147,106 @@ func (s *Server) initMetrics(ctx context.Context, installCfg config.Install) (rR
+// value change can be observed. Currently, this is done for "meter", "timer" and "histogram" types.
+func isZeroValueMetric(metricType string) bool {
+	return metricType == "meter" || metricType == "timer" || metricType == "histogram"
+}
+
+// zeroValuesForMetricType returns the metric values for the "zero" value of the metric of the given type. The provided
+// metric type must be a valid type that returns true when provided to the "isZeroValueMetric" function: panics
+// otherwise.
+func zeroValuesForMetricType(metricType string) map[string]interface{} {
+	var zeroMetric interface{}
+	switch metricType {
+	case "meter":
+		zeroMetric = gometrics.NewMeter()
+	case "timer":
+		zeroMetric = gometrics.NewTimer()
+	case "histogram":
+		zeroMetric = gometrics.NewHistogram(metrics.DefaultSample())
+	default:
+		panic(fmt.Sprintf("should never happen: unsupported metricType %s", metricType))
+	}
+	return metrics.ToMetricVal(zeroMetric).Values()
+}
+
+// logZeroValueMetric logs a "zero" value for metric with the provided metricID, metricType, and tags to the provided
+// metricLogger if an entry for the metric has not yet been logged (which is determined based on whether or not an entry
+// for the metric exists in the provided seenMetrics set). Returns the zero values of the logged metric. If a zero value
+// is logged, an entry for the metric is added to seenMetrics.
+func logZeroValueMetric(
+	metricLogger metric1log.Logger,
+	metricID,
+	metricType string,
+	tags map[string]string,
+	seenMetrics *seenMetricsSet,
+	metricTypeValuesDisallowedList map[string]map[string]struct{},
+	metricTagsParam metric1log.Param,
+) (zeroValuesLogged map[string]interface{}) {
+
+	// if metric has been seen before, no need to log zero value
+	mapKey := tagMapKey(metricID, metricType, tags)
+
+	// only acquire read lock while checking because this should be true vast majority of the time
+	seenMetrics.lock.RLock()
+	_, metricSeen := seenMetrics.seenSet[mapKey]
+	seenMetrics.lock.RUnlock()
+	if metricSeen {
+		return nil
+	}
+
+	zeroValuesToUse := zeroValuesForMetricType(metricType)
+	removeDisallowedKeys(metricType, zeroValuesToUse, metricTypeValuesDisallowedList)
+
+	// if removing the disallowed keys causes the values to be empty, no need to log zero value
+	if len(zeroValuesToUse) == 0 {
+		return nil
+	}
+
+	seenMetrics.lock.Lock()
+	// need to check again while under lock because read lock was only held during initial check to optimize for fact
+	// that it will return true vast majority of the time. However, want to ensure that only one zero-value log line is
+	// emitted, so check condition again while holding write lock.
+	if _, ok := seenMetrics.seenSet[mapKey]; !ok {
+		// metric not seen before: emit zero-value and record
+		metricLogger.Metric(metricID, metricType, metric1log.Values(zeroValuesToUse), metricTagsParam)
+		seenMetrics.seenSet[mapKey] = struct{}{}
+		zeroValuesLogged = zeroValuesToUse
+	}
+	seenMetrics.lock.Unlock()
+	return zeroValuesLogged
```

Comment:
Thanks -- was mainly matching the style of the read lock, which was done on its own since we didn't want to defer in that case.

However, upon thinking about this further, I think that I'm going to update the implementation to just acquire a write lock. I believe that 99%+ of the time, this function will only be called in a single goroutine (the "emit" routine) -- the only instance where this isn't the case is when "emitFn" is called on shutdown. As such, I think it's cheaper to just acquire and release the lock once even if it's over a bigger section of logic -- the current optimization only helps when there may be more reads than writes from different goroutines, but since that isn't the case here I think the simplification will be worth it.

Reply:
re style: I figured that was the case, but usually still fall back to the `defer` use case if possible later since the explicit locking usually only applies in one-off scenarios.

That was the reason I posed the question in the draft initially. It sounds like there will be very little lock contention (except for on shutdown), which as you eluded to makes more sense to just lock over all of the logic. I am in sync on this and it cleans up the code path nicely.